### DISCUSSION TITLE
3.5 WF to Elsa workflow

### DISCUSSION THREAD

**Discussion by nikunj231991** [elsa-workflows/elsa-core#2522] [Design & tech]
We have existing project which is working with .net framework WF 3.5.

We want to migrate .net framework to .net 5/6 but due to WF 3.5 we can not.

So can we port WF 3.5 to Elsa workflow?

**Comment by sfmskywalker** [comment]
Hi @nikunj231991 ,

Makes sense.

Whether you can port over the WF 3.5 solution to Elsa largely depends on what features you use in WF and whether there is an equivalent for it in Elsa.

**Reply by nikunj231991** [reply]
We have used below features from WF3.5
1) Workflow Events
2) Delay Events
3) State machine workflow
4) WorkflowServiceHost 
5) Data Contract Resolver
6) SQL Workflow Instance Store
7) Custom activities

Does this all supported by Elsa?

**Reply by sfmskywalker** [reply]
- [x] 1. Workflow Events
- [x] 2. Delay Events
- [x] 3. State machine workflow *
- [x] 4. WorkflowServiceHost *
- [x] 5. Data Contract Resolver *
- [x] 6. SQL Workflow Instance Store
- [x] 7. Custom activities

In that case there should be no problem. However, the **state machine workflow* model is not natively supported by Elsa in the same way of WF.

In Elsa, you implement the State machine like a regular flowchart workflow, but you use the `State` activity to set the current state.
To transition between states, you would add the necessary triggers.

Please see this [SO answer](https://stackoverflow.com/a/67858267/690374) for more details.

With regards to the other marked items, they don't exists "as is" in Elsa, but there are equivalent services and options.

**Reply by clareychef** [reply]
I've tried implementing the state activity, but it's not working quite right for me. There is no Transition option, for starters. I'm using Elsa 2.5. Any suggestions please?

**Reply by sfmskywalker** [reply]
Correct, Elsa 2 wasn't designed with native state machine primitives unfortunately. To model a "transition", you would do so using triggers.

For example, let's say _State 1_ has two three outcomes: _Transition 1_, _Transition 2_ and _Transition 3_, and you want to transition to _State 2_ via `_Transition 2` when something happens.

To implement that, you need to decide what "something" means. Maybe it's a signal sent from your application. In which case, you would add the **Signal Received** activity between _State 1_ and _State 2_ via the _Transition 2_ outcome of _State 1_.

For Elsa 3, the state machine model will be designed to be that of a more traditional state machine.
